hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeUpgradeActions.java

```diff
@@ -0,0 +1,196 @@
+        requireNonNull(adminServiceConfig);
+        requireNonNull(freezeStore);
+
+        this.adminServiceConfig = adminServiceConfig;
+        this.freezeStore = freezeStore;
+    }
+
+    public void externalizeFreezeIfUpgradePending() {
+        // TODO: call networkCtx.hasPreparedUpgrade() once available (issue #6201)
+        // final var isUpgradePrepared = networkCtx.hasPreparedUpgrade();
+        final boolean isUpgradePrepared = true;
+
+        if (isUpgradePrepared) {
+            writeCheckMarker(NOW_FROZEN_MARKER);
+        }
+    }
+
+    public CompletableFuture<Void> extractTelemetryUpgrade(
+            @NonNull final byte[] archiveData, @Nullable final Instant now) {
+        requireNonNull(archiveData);
+        return extractNow(archiveData, TELEMETRY_UPGRADE_DESC, EXEC_TELEMETRY_MARKER, now);
+    }
+
+    public CompletableFuture<Void> extractSoftwareUpgrade(@NonNull final byte[] archiveData) {
+        requireNonNull(archiveData);
+        return extractNow(archiveData, PREPARE_UPGRADE_DESC, EXEC_IMMEDIATE_MARKER, null);
+    }
+
+    public void scheduleFreezeOnlyAt(@NonNull final Instant freezeTime) {
+        requireNonNull(freezeTime);
+        withNonNullDualState("schedule freeze", ds -> ds.freezeTime(freezeTime));
+    }
+
+    public void scheduleFreezeUpgradeAt(@NonNull final Instant freezeTime) {
+        requireNonNull(freezeTime);
+        withNonNullDualState("schedule freeze", ds -> {
+            ds.freezeTime(freezeTime);
+            writeSecondMarker(FREEZE_SCHEDULED_MARKER, freezeTime);
+        });
+    }
+
+    public void abortScheduledFreeze() {
+        withNonNullDualState("abort freeze", ds -> {
+            ds.freezeTime(null);
+            writeCheckMarker(FREEZE_ABORTED_MARKER);
+        });
+    }
+
+    public boolean isFreezeScheduled() {
+        final var ans = new AtomicBoolean();
+        withNonNullDualState("check freeze schedule", ds -> {
+            final var freezeTime = ds.freezeTime();
+            ans.set(freezeTime != null && !freezeTime.equals(ds.lastFrozenTime()));
+        });
+        return ans.get();
+    }
+
+    /* --- Internal methods --- */
+
+    private CompletableFuture<Void> extractNow(
+            @NonNull final byte[] archiveData,
+            @NonNull final String desc,
+            @NonNull final String marker,
+            @Nullable final Instant now) {
+        requireNonNull(archiveData);
+        requireNonNull(desc);
+        requireNonNull(marker);
+
+        final int size = archiveData.length;
+        final String artifactsLoc = adminServiceConfig.upgradeArtifactsPath();
+        requireNonNull(artifactsLoc);
+        log.info("About to unzip {} bytes for {} update into {}", size, desc, artifactsLoc);
+        // we spin off a separate thread to avoid blocking handleTransaction
+        // if we block handle, there could be a dramatic spike in E2E latency at the time of PREPARE_UPGRADE
+        return runAsync(() -> {
```

Comment:
This is enough code that it probably should be extracted to a private method, rather than an inline lambda.
something like `runAsync(this::extractAndReplaceArtifacts)` could then be used to reduce side effects and avoid generating additional capture code.
It it's strictly necessary to capture the local variables, at least you could write the clearer `runAsync(() extractAndReplaceArtifacts(archiveData, marker, now, desc))` (assuming the size and artifactsLoc calculations are moved inside the new method)

Reply:
Fixed